[SPARK-42169] [SQL] Implement code generation for to_csv function (StructsToCsv)

NarekDW opened the pull request:
### What changes were proposed in this pull request?
This PR enhances `StructsToCsv` class with `doGenCode` function instead of extending it from `CodegenFallback` trait (performance improvement).

### Why are the changes needed?
It will improve performance.

### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
an additional test case were added to `org.apache.spark.sql.CsvFunctionsSuite` class.

NarekDW committed:
make to_csv function deterministic

NarekDW committed:
remove redundant Serializable definition from StructsToCsv case class

NarekDW committed:
remove redundant test case

NarekDW committed:
add test case to check generated code doesn't exceed the limit by size

MaxGekk reviewed (commented):
Does `CSVBenchmark` show any performance improvements?

NarekDW commented:
@MaxGekk 
This is from master branch:

<img width="981" alt="image" src="https://user-images.githubusercontent.com/24260474/216346895-1b25683c-186a-47ce-be50-e912255f061e.png">

This is from current branch:

<img width="984" alt="image" src="https://user-images.githubusercontent.com/24260474/216347344-14bbb4ee-0915-4228-b2a6-f032ee4bd909.png">

On my local machine - `to_csv` related operations in average are about ~20 % faster with this change in `CSVBenchmark`.

MaxGekk commented:
@NarekDW Could you regenerate benchmark results using GitHub actions, see
https://spark.apache.org/developer-tools.html (Running benchmarks in your forked repository)
and update:
- CSVBenchmark-results.txt
- CSVBenchmark-jdk11-results.txt
- CSVBenchmark-jdk17-results.txt

in your PR.

NarekDW commented:
> @NarekDW Could you regenerate benchmark results using GitHub actions, see https://spark.apache.org/developer-tools.html (Running benchmarks in your forked repository) and update:
> 
> * CSVBenchmark-results.txt
> * CSVBenchmark-jdk11-results.txt
> * CSVBenchmark-jdk17-results.txt
> 
> in your PR.

@MaxGekk sorry for the late response. I've added benchmark results from GitHub actions.
I've added 2 commits: in pre-last commit I've updated benchmarks results from **master** branch and in the last commit I've added  benchmarks results from **current** branch.

Execution links for master branch:
[Java 8](https://github.com/NarekDW/spark/actions/runs/4151578967)
[Java 11](https://github.com/NarekDW/spark/actions/runs/4151828105)
[Java 17](https://github.com/NarekDW/spark/actions/runs/4152154541)

Execution links for current branch:
[Java 8](https://github.com/NarekDW/spark/actions/runs/4152491256)
[Java 11](https://github.com/NarekDW/spark/actions/runs/4152723531)
[Java 17](https://github.com/NarekDW/spark/actions/runs/4155468338)

P.S. **Scala 2.12** was used for all executions.

jaceklaskowski commented on sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:
```diff
@@ -574,4 +575,11 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
+
+  test("StructsToCsv should not generate codes beyond 64KB") {
+    val range = Range.inclusive(1, 5000)
+    val struct = CreateStruct.create(range.map(Literal(_)))
```
nit: `Literal.apply`

NarekDW replied:
done

NarekDW committed:
minor scala style fix

NarekDW commented:
@jaceklaskowski thank you for the review. @MaxGekk just a reminder.

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:
```diff
@@ -577,4 +578,11 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
     checkAnswer(actual, Row(Row(1, "2\n2")))
   }
+
+  test("StructsToCsv should not generate codes beyond 64KB") {
```
Could you clarify this test title, please. I don't see anything related to 64KB in the test.

Doesn't the test just duplicates existing one: `to_csv - struct`?

NarekDW replied:
Sure, `checkEvaluation` will execute this test in both modes without codegen and with codegen modes. In case of codegen mode, if the generated java code (method) will be larger than 64 kb in size it won't be able to compile it, and the test will fail, as Java has a 64kb limit on the size of methods.
It doesn't duplicate `to_csv - struct` test case, because the purpose of this test - is to generate big StructType(with 5000 literals in current case) and test it on `StructsToCsv` expression to be sure that generated java code doesn't exceed the limit of 64 kb in size.

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:
```diff
@@ -577,4 +578,11 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
+    val range = Range.inclusive(1, 5000)
+    val struct = CreateStruct.create(range.map(Literal.apply))
+    val expected = range.mkString(",")
+    checkEvaluation(StructsToCsv(Map.empty, struct), expected)
```
It is better to place the test to `CsvExpressionsSuite` because `CsvFunctionsSuite` aims to check functions like `to_csv`/`from_csv`.

NarekDW replied:
Yep, done.

NarekDW committed:
move test case from CsvFunctionsSuite to CsvExpressionsSuite

MaxGekk commented:
@NarekDW Could you rebase this PR on the recent master, please.

NarekDW commented:
@MaxGekk sure, I've rebased, but there were conflicts in benchmarks. It will take some time to regenerate them.

MaxGekk reviewed (commented):
LGTM, waiting for benchmark results.

NarekDW committed:
update benchmarks

NarekDW commented:
@MaxGekk sorry for delay, benchmark results are updated.

MaxGekk commented:
+1, LGTM. Merging to master.
Thank you, @NarekDW and @HyukjinKwon @jaceklaskowski for review.
